### PR TITLE
Add SAML IdP service providers to the cache and CLI.

### PR DESCRIPTION
**Pull request by mdwn**
SAML IdP service providers have been added to the cache and are now manageable via the CLI. Some alterations to role permissions have occurred to ensure that the system and users can manage these.

### CODE REVIEW

File: lib/cache/collections.go [hunk 1]
```diff
@@ -2686,3 +2691,95 @@ func (s *kubeCluster) processEvent(ctx context.Context, event types.Event) error
...
+	if err := s.samlIdpServiceProvidersCache.DeleteAllSAMLIdPServiceProviders(ctx); err != nil {
+		if !trace.IsNotFound(err) {
+			return trace.Wrap(err)
+		}
+	}
+	return nil
+}
+
+func (s *samlIDPServiceProviders) fetch(ctx context.Context) (apply func(ctx context.Context) error, err error) {
+	resources := make([]types.SAMLIdPServiceProvider, 0)
```
Comment (tigrato) [inline lib/cache/collections.go:2710]: why make?

File: lib/cache/collections.go [hunk 2]
```diff
@@ -2686,3 +2691,95 @@ func (s *kubeCluster) processEvent(ctx context.Context, event types.Event) error
...
+
+		for _, resource := range resources {
+			if _, err := s.samlIdpServiceProvidersCache.GetSAMLIdPServiceProvider(ctx, resource.GetName()); err != nil {
+				if trace.IsNotFound(err) {
+					err = s.samlIdpServiceProvidersCache.CreateSAMLIdPServiceProvider(ctx, resource)
+					if err != nil {
+						return trace.Wrap(err)
+					}
+				}
+			}
```
Comment (tigrato) [inline lib/cache/collections.go:2740]:
```suggestion
			if _, err := s.samlIdpServiceProvidersCache.GetSAMLIdPServiceProvider(ctx, resource.GetName()); trace.IsNotFound(err) {
					err = s.samlIdpServiceProvidersCache.CreateSAMLIdPServiceProvider(ctx, resource)
					if err != nil {
						return trace.Wrap(err)
					}
				
			}else if err!=nil{
			
			return trace.Wrap(err)
			}
```

Should we upsert when we create the resource or should we just create and return?

File: lib/cache/collections.go [hunk 3]
```diff
@@ -2686,3 +2691,95 @@ func (s *kubeCluster) processEvent(ctx context.Context, event types.Event) error
...
+func (s *samlIDPServiceProviders) processEvent(ctx context.Context, event types.Event) error {
+	switch event.Type {
+	case types.OpDelete:
+		err := s.samlIdpServiceProvidersCache.DeleteSAMLIdPServiceProvider(ctx, event.Resource.GetName())
+		if err != nil {
+			if !trace.IsNotFound(err) {
+				s.Logger.WithError(err).Warn("Failed to delete resource.")
+				return trace.Wrap(err)
+			}
+		}
```
Comment (tigrato) [inline lib/cache/collections.go:2758]:
```suggestion
		if err != nil && !trace.IsNotFound(err){

				s.Logger.WithError(err).Warn("Failed to delete resource.")
				return trace.Wrap(err)
		
		}
```

File: lib/cache/collections.go [hunk 4]
```diff
@@ -2686,3 +2691,95 @@ func (s *kubeCluster) processEvent(ctx context.Context, event types.Event) error
...
+	case types.OpPut:
+		resource, ok := event.Resource.(types.SAMLIdPServiceProvider)
+		if !ok {
+			return trace.BadParameter("unexpected type %T", event.Resource)
+		}
+		if _, err := s.samlIdpServiceProvidersCache.GetSAMLIdPServiceProvider(ctx, resource.GetName()); err != nil {
+			if trace.IsNotFound(err) {
```
Comment (tigrato) [inline lib/cache/collections.go:2765]: Same suggestions as above

File: tool/tctl/common/resource_command.go [hunk 5]
```diff
@@ -1520,6 +1558,30 @@ func (rc *ResourceCommand) getCollection(ctx context.Context, client auth.Client
 			Name: rc.ref.Name,
 		})
 		return &loginRuleCollection{[]*loginrulepb.LoginRule{rule}}, trail.FromGRPC(err)
+	case types.KindSAMLIdPServiceProvider:
+		if rc.ref.Name == "" {
+			resources := make([]types.SAMLIdPServiceProvider, 0)
```
Comment (tigrato) [inline tool/tctl/common/resource_command.go:1563]: avoid make on empty arrays

File: lib/cache/collections.go (hunk 1 above)
Comment (mdwn) [reply lib/cache/collections.go:2710]: Bad habit!

File: lib/cache/collections.go (hunk 2 above)
Comment (mdwn) [reply lib/cache/collections.go:2740]: Ahh whoops, you're right! I'll return after the create.

File: lib/cache/collections.go [hunk 6]
```diff
@@ -2686,3 +2691,94 @@ func (s *kubeCluster) processEvent(ctx context.Context, event types.Event) error
...
+}
+
+func (s *samlIDPServiceProviders) fetch(ctx context.Context) (apply func(ctx context.Context) error, err error) {
+	var resources []types.SAMLIdPServiceProvider
+
+	nextKey := ""
+	for {
+		var samlProviders []types.SAMLIdPServiceProvider
+		var err error
+		samlProviders, nextKey, err = s.SAMLIdPServiceProviders.ListSAMLIdPServiceProviders(ctx, 200, nextKey)
```
[the remark this answers is not in the document]
Comment (mdwn) [reply lib/cache/collections.go:2716]: I'll use 0 here, that makes sense.

File: lib/cache/collections.go [hunk 7]
```diff
@@ -2686,3 +2691,94 @@ func (s *kubeCluster) processEvent(ctx context.Context, event types.Event) error
...
+		}
+	}
+
+	return func(ctx context.Context) error {
+		if err := s.erase(ctx); err != nil {
+			return trace.Wrap(err)
+		}
+
+		for _, resource := range resources {
+			if _, err := s.samlIdpServiceProvidersCache.GetSAMLIdPServiceProvider(ctx, resource.GetName()); trace.IsNotFound(err) {
```
[the remark this answers is not in the document]
Comment (mdwn) [reply lib/cache/collections.go:2733]: Ah, interesting. I don't think there's any benefit over that method. I'll go ahead and use that instead.

File: lib/cache/cache.go [hunk 8]
```diff
@@ -148,6 +149,7 @@ func ForProxy(cfg Config) Config {
 		{Kind: types.KindKubeServer},
 		{Kind: types.KindInstaller},
 		{Kind: types.KindKubernetesCluster},
+		{Kind: types.KindSAMLIdPServiceProvider},
```
[the remark this answers is not in the document]
Comment (mdwn) [reply lib/cache/cache.go:152]: Hrm, Is the implication here that we need to get this in before 12 is released if I want this to be cached?

[the remark this answers is not in the document]
Comment (r0mant) [reply lib/cache/cache.go:152]: Can proxy use a similar version check with auth server? Or do you have other suggestions?

[the remark this answers is not in the document]
Comment (mdwn) [reply lib/cache/cache.go:152]: Okay, so I think the decision for now is that, when this is backported to v12, this change to `ForProxy` will be removed.

**Review by rosstimothy — APPROVED** [review]

**Review by tigrato — APPROVED** [review]

**Pull request merged**